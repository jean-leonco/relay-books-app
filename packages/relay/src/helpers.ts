import { CacheConfig, RequestParameters, Variables } from 'relay-runtime';

export const isMutation = (request: RequestParameters) => request.operationKind === 'mutation';
export const isQuery = (request: RequestParameters) => request.operationKind === 'query';
export const forceFetch = (cacheConfig: CacheConfig) => !!(cacheConfig && cacheConfig.force);

export const handleData = (response) => {
  const contentType = response.headers.get('content-type');
  if (contentType && contentType.indexOf('application/json') !== -1) {
    return response.json();
  }

  return response.text();
};

export const getRequestBody = (request: RequestParameters, variables: Variables) => {
  return JSON.stringify({
    name: request.name,
    query: request.text, // GraphQL text from input
    variables,
  });
};
