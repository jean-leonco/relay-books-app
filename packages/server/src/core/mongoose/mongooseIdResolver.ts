import { GraphQLNonNull, GraphQLString } from 'graphql';
import { Types } from 'mongoose';

type MongooseModel = {
  _id: Types.ObjectId;
};

export const mongooseIdResolver = {
  _id: {
    type: GraphQLNonNull(GraphQLString),
    description: 'MongoDB _id',
    resolve: ({ _id }: MongooseModel) => _id.toString(),
  },
};
