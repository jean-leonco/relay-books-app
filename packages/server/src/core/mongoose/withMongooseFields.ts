export const isActiveMongooseField = {
  isActive: {
    type: Boolean,
    description: 'Soft delete. Usually when users want to hide something from public view.',
    required: true,
    default: true,
    index: true,
  },
};

export const removedAtMongooseField = {
  removedAt: {
    type: Date,
    description: 'Hard delete',
    default: null,
  },
};
