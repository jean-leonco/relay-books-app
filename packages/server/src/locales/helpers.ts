import i18next, { TOptions } from 'i18next';

import { NamespaceKeys, MessageKeys } from './en';

export const t = (namespace: NamespaceKeys, message: MessageKeys, options?: TOptions) =>
  i18next.t(`${namespace}::${message}`, options);
