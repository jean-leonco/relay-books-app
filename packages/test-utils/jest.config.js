const pkg = require('./package');

module.exports = {
  rootDir: './',
  name: pkg.name,
  displayName: pkg.name.toUpperCase(),
  testPathIgnorePatterns: ['/node_modules/', './build'],
  coverageReporters: ['lcov', 'html'],
  setupFilesAfterEnv: ['<rootDir>/test/setupTestFramework.js'],
  globalSetup: '<rootDir>/test/setup.js',
  globalTeardown: '<rootDir>/test/teardown.js',
  transform: {
    // '^.+\\.(js|jsx|ts|tsx)?$': require.resolve('babel-jest'),
    '^.+\\.(js|jsx|ts|tsx)?$': '<rootDir>/test/babel-transformer',
  },
};
