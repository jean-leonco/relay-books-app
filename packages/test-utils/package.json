{
  "name": "@booksapp/test-utils",
  "version": "0.0.1",
  "private": true,
  "license": "UNLICENSED",
  "main": "src/index.ts",
  "module": "src/index.ts",
  "scripts": {
    "test": "jest",
    "test:watch": "jest --watch"
  },
  "dependencies": {
    "@booksapp/babel": "^0.0.1",
    "@booksapp/types": "^0.0.1",
    "@koa/router": "^9.3.1",
    "babel-jest": "^26.1.0",
    "graphql-relay": "0.6.0",
    "koa": "^2.13.0",
    "koa-bodyparser": "^4.3.0",
    "mongoose": "^5.10.7",
    "supertest": "^5.0.0"
  }
}
