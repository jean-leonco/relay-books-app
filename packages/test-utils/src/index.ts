export { defaultFrozenKeys, sanitizeValue, sanitizeTestObject } from './objectNormalization';
export { httpRequestGraphql } from './httpRequestGraphql';
