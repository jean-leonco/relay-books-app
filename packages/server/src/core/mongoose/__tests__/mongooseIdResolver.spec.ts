import mongoose, { Document, Model } from 'mongoose';
import {
  graphql,
  GraphQLBoolean,
  GraphQLObjectType,
  GraphQLSchema,
  GraphQLString,
  GraphQLNonNull,
  GraphQLID,
} from 'graphql';
import { toGlobalId, fromGlobalId, globalIdField } from 'graphql-relay';

import { sanitizeTestObject } from '@booksapp/test-utils';

import { clearDbAndRestartCounters, connectMongoose, disconnectMongoose, gql } from '../../../../test/helpers';

import { mongooseIdResolver } from '../mongooseIdResolver';

beforeAll(connectMongoose);

beforeEach(clearDbAndRestartCounters);

afterAll(disconnectMongoose);

const Schema = new mongoose.Schema(
  {
    field: {
      type: String,
      description: 'This is a string field',
      required: true,
    },
  },
  {
    collection: 'MyModel',
  },
);

export interface IMyModel extends Document {
  field: string;
}

const MyModel: Model<IMyModel> = mongoose.model('MyModel', Schema);

const MyType = new GraphQLObjectType<IMyModel>({
  name: 'MyType',
  fields: () => ({
    id: globalIdField('MyType'),
    ...mongooseIdResolver,
    field: {
      type: GraphQLString,
      description: 'This is a string field',
      resolve: (obj) => obj.field,
    },
  }),
});

const query = new GraphQLObjectType({
  name: 'Query',
  fields: () => ({
    _: {
      type: GraphQLBoolean,
      resolve: () => true,
    },
    myType: {
      type: MyType,
      args: {
        id: {
          type: GraphQLNonNull(GraphQLID),
        },
      },
      resolve: async (_, args) => await MyModel.findById(fromGlobalId(args.id).id),
    },
  }),
});

const schema = new GraphQLSchema({
  // query is obligatory
  query,
});

describe('Mongoose Id Resolver', () => {
  it('should get myType with mongooseIdResolver', async () => {
    const myModel = await new MyModel({ field: 'value' }).save();

    const query = gql`
      query Q($id: ID!) {
        myType(id: $id) {
          id
          _id
          field
        }
      }
    `;

    const rootValue = {};
    const variables = { id: toGlobalId('MyType', myModel._id) };
    const result = await graphql(schema, query, rootValue, {}, variables);

    expect(result.errors).toBeUndefined();
    expect(result.data?.myType).not.toBe(null);
    expect(result.data?.myType._id).toEqual(myModel._id.toString());
    expect(sanitizeTestObject(result.data)).toMatchSnapshot();
  });
});
