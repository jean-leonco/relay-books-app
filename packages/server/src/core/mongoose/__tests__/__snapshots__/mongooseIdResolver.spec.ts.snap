// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Mongoose Id Resolver should get myType with mongooseIdResolver 1`] = `
Object {
  "myType": Object {
    "_id": "ObjectId",
    "field": "value",
    "id": "FROZEN-ID",
  },
}
`;
